File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    time = 0
    # 대기중인 트럭
    truck_weights = deque(truck_weights)
    # 다리
    b = [0 for _ in range(bridge_length)]
    b = deque(b)

    current_weight = 0 
    
    while truck_weights :
        time += 1
        current_weight -= b.popleft()
        if current_weight + truck_weights[0] <= weight :
            current_weight += truck_weights[0]
            b.append(truck_weights.popleft())
        else :
            b.append(0)
    time += bridge_length
    return time

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
    var answer = 0;
    
    // 다리
    let queue = new Array(bridge_length).fill(0);
    
    // 다리 무게
    let bridge_weight = 0;
    
    while(queue.length) {
        let q = queue.shift();
        bridge_weight -= q;
        let pop = truck_weights[0];
        
        if(truck_weights.length) {
            // 다리 최대 무게보다 가벼우면
            if(bridge_weight + pop <= weight) {
                let truck = truck_weights.shift();
                queue.push(truck);
                bridge_weight += pop;
            }
            // 다리 최대 무게보다 무거우면
            else {
                queue.push(0);
            }
        }
        answer += 1;
    }
    return answer;
}